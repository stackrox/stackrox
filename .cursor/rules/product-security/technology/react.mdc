---
description: Secure coding rules for React projects ⚛️
globs:
  - "**/*.jsx"
  - "**/*.tsx"
alwaysApply: false
---

- Use default React data binding to prevent XSS.
- Sanitize HTML before using dangerouslySetInnerHTML.
- Validate URLs before using them in the DOM.
- Secure server-side rendering and avoid JSON injection.
