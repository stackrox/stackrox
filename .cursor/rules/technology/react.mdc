# React-specific secure coding rules for Cursor
# This file provides security best practices for React projects.
---
description: Secure coding rules for React projects ⚛️
globs:
  - "**/*.jsx"
  - "**/*.tsx"
  - "**/src/**/*.jsx"
  - "**/src/**/*.tsx"
  - "**/components/**/*.jsx"
  - "**/components/**/*.tsx"
  - "**/pages/**/*.jsx"
  - "**/pages/**/*.tsx"
  - "**/App.jsx"
  - "**/App.tsx"
  - "**/index.jsx"
  - "**/index.tsx"
alwaysApply: false
---

- Use default React data binding to prevent XSS.
- Sanitize HTML before using dangerouslySetInnerHTML.
- Validate URLs before using them in the DOM.
- Secure server-side rendering and avoid JSON injection.
