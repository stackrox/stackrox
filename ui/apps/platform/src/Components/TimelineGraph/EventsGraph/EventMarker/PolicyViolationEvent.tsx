import { forwardRef } from 'react';

type PolicyViolationEventProps = {
    size: number;
};

const PolicyViolationEvent = forwardRef<SVGSVGElement, PolicyViolationEventProps>(
    ({ size }, ref) => {
        return (
            <svg
                data-testid="policy-violation-event"
                width={size}
                height={size}
                viewBox="0 0 15 15"
                version="1.1"
                xmlns="http://www.w3.org/2000/svg"
                ref={ref}
            >
                <g transform="translate(0 .1)" fill="none">
                    <rect fill="#FF5782" width="14.58" height="14.58" rx="2.43" />
                    <path
                        d="M8.071 8.37H6.616l-.325-5.584h2.12L8.07 8.37zm-1.833 2.22c0-.346.091-.614.275-.804.184-.19.458-.286.82-.286.364 0 .639.094.825.282.186.189.279.458.279.808 0 .345-.096.613-.29.804-.192.19-.463.285-.813.285-.359 0-.631-.096-.817-.289-.186-.192-.28-.46-.28-.8z"
                        fill="#FFF"
                    />
                </g>
            </svg>
        );
    }
);

PolicyViolationEvent.displayName = 'PolicyViolationEvent';

export default PolicyViolationEvent;
