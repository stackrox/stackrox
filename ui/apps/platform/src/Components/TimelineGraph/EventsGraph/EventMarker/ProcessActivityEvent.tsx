import { forwardRef } from 'react';

type ProcessActivityEventProps = {
    inBaseline?: boolean;
    size: number;
};

const ProcessActivityEvent = forwardRef<SVGSVGElement, ProcessActivityEventProps>(
    ({ inBaseline = false, size }, ref) => {
        return inBaseline ? (
            <svg
                data-testid="process-in-baseline-activity-event"
                width={size}
                height={size}
                viewBox="0 0 16 16"
                xmlns="http://www.w3.org/2000/svg"
                ref={ref}
            >
                <g transform="translate(0 .1)" fill="none" fillRule="evenodd">
                    <rect fill="#56DDB2" width="14.58" height="14.58" rx="2.43" />
                    <path
                        d="M4.459 6.768a.807.807 0 00-1.18-.072.91.91 0 00-.067 1.24l2.685 3.17a.81.81 0 001.281-.043l4.645-6.244a.909.909 0 00-.144-1.233.808.808 0 00-1.172.151L6.48 9.153l-2.02-2.385z"
                        fill="#FFF"
                        fillRule="nonzero"
                    />
                </g>
            </svg>
        ) : (
            <svg
                data-testid="process-activity-event"
                width={size}
                height={size}
                viewBox="0 0 16 16"
                xmlns="http://www.w3.org/2000/svg"
                ref={ref}
            >
                <rect
                    x="689"
                    y="673.1"
                    width="14.58"
                    height="14.58"
                    rx="2.43"
                    transform="translate(-689 -673)"
                    fill="#5677DD"
                    fillRule="evenodd"
                />
            </svg>
        );
    }
);

ProcessActivityEvent.displayName = 'ProcessActivityEvent';

export default ProcessActivityEvent;
