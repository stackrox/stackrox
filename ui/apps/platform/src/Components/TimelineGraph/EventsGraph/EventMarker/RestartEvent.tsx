import { forwardRef } from 'react';

type RestartEventProps = {
    size: number;
};

const RestartEvent = forwardRef<SVGSVGElement, RestartEventProps>(({ size }, ref) => {
    return (
        <svg
            data-testid="restart-event"
            width={size}
            height={size}
            viewBox="0 0 16 15"
            xmlns="http://www.w3.org/2000/svg"
            ref={ref}
            fillRule="evenodd"
            clipRule="evenodd"
            strokeLinejoin="round"
            strokeMiterlimit="2"
        >
            <path
                d="M9.202 1.183l6.612 11.268a1.354 1.354 0 01-.508 1.866c-.211.12-.451.183-.694.183H1.388A1.377 1.377 0 010 13.134c0-.24.064-.475.186-.683L6.798 1.183a1.4 1.4 0 012.404 0z"
                fill="#ff9064"
            />
        </svg>
    );
});

RestartEvent.displayName = 'RestartEvent';

export default RestartEvent;
