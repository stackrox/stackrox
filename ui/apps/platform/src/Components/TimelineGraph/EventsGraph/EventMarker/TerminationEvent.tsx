import { forwardRef } from 'react';

type TerminationEventProps = {
    size: number;
};

const TerminationEvent = forwardRef<SVGSVGElement, TerminationEventProps>(({ size }, ref) => {
    return (
        <svg
            data-testid="termination-event"
            width={size}
            height={size}
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
            ref={ref}
        >
            <path
                d="M9.202 13.817l6.612-11.268a1.354 1.354 0 00-.508-1.866A1.406 1.406 0 0014.612.5H1.388C.621.5 0 1.112 0 1.866c0 .24.064.475.186.683l6.612 11.268a1.4 1.4 0 002.404 0z"
                fill="#FF9064"
                fillRule="evenodd"
            />
        </svg>
    );
});

TerminationEvent.displayName = 'TerminationEvent';

export default TerminationEvent;
