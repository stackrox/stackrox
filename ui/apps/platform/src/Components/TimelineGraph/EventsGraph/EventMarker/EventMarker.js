import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import { scaleLinear } from 'd3-scale';

import { getWidth } from 'utils/d3Utils';
import { eventTypes } from 'constants/timelineTypes';
import mainViewSelector from 'Components/TimelineGraph/MainView/selectors';
import D3Anchor from 'Components/D3Anchor';
import EventTooltip from 'Components/TimelineGraph/EventsGraph/EventTooltip';
import PolicyViolationEvent from './PolicyViolationEvent';
import RestartEvent from './RestartEvent';
import ProcessActivityEvent from './ProcessActivityEvent';
import TerminationEvent from './TerminationEvent';

const EventMarker = ({
    id,
    name,
    args,
    type,
    uid,
    parentName,
    parentUid,
    reason,
    timestamp,
    inBaseline,
    differenceInMilliseconds,
    translateX,
    translateY,
    minTimeRange,
    maxTimeRange,
    size,
    margin,
}) => {
    const popoverRef = useRef(null);

    // the "container" argument is a reference to the container for the D3-related element
    function onUpdate(container) {
        const width = getWidth(mainViewSelector);
        const minRange = margin;
        const maxRange = width - margin;
        const xScale = scaleLinear()
            .domain([minTimeRange, maxTimeRange])
            .range([minRange, maxRange]);
        const x = xScale(differenceInMilliseconds).toFixed(0);

        container.attr(
            'transform',
            `translate(${Number(translateX) + Number(x) - size / 2}, ${
                Number(translateY) - size / 2
            })`
        );
    }

    return (
        <D3Anchor
            dataTestId="timeline-event-marker"
            translateX={translateX}
            translateY={translateY}
            onUpdate={onUpdate}
        >
            <EventTooltip
                id={id}
                name={name}
                args={args}
                type={type}
                uid={uid}
                parentName={parentName}
                parentUid={parentUid}
                timestamp={timestamp}
                reason={reason}
                inBaseline={inBaseline}
                popoverRef={popoverRef}
            >
                <g>
                    {type === eventTypes.POLICY_VIOLATION && (
                        <PolicyViolationEvent size={size} ref={popoverRef} />
                    )}
                    {type === eventTypes.PROCESS_ACTIVITY && (
                        <ProcessActivityEvent
                            size={size}
                            inBaseline={inBaseline}
                            ref={popoverRef}
                        />
                    )}
                    {type === eventTypes.RESTART && <RestartEvent size={size} ref={popoverRef} />}
                    {type === eventTypes.TERMINATION && (
                        <TerminationEvent size={size} ref={popoverRef} />
                    )}
                </g>
            </EventTooltip>
        </D3Anchor>
    );
};

EventMarker.propTypes = {
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    args: PropTypes.string,
    type: PropTypes.string.isRequired,
    uid: PropTypes.number,
    parentName: PropTypes.string,
    parentUid: PropTypes.number,
    reason: PropTypes.string,
    timestamp: PropTypes.string.isRequired,
    inBaseline: PropTypes.bool,
    differenceInMilliseconds: PropTypes.number.isRequired,
    translateX: PropTypes.number.isRequired,
    translateY: PropTypes.number.isRequired,
    minTimeRange: PropTypes.number.isRequired,
    maxTimeRange: PropTypes.number.isRequired,
    size: PropTypes.number.isRequired,
    margin: PropTypes.number,
};

EventMarker.defaultProps = {
    uid: null,
    parentName: null,
    parentUid: null,
    args: null,
    reason: null,
    inBaseline: false,
    margin: 0,
};

export default EventMarker;
