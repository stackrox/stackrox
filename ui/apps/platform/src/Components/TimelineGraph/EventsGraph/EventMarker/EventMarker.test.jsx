import React from 'react';
import { render, screen } from '@testing-library/react';

import EventMarker from './EventMarker';

test('should show a policy violation event marker', async () => {
    const { asFragment } = render(
        <svg height={100} width={100} data-testid="timeline-main-view">
            <EventMarker
                id="1"
                name="eventName"
                type="PolicyViolationEvent"
                timestamp="2020-04-20T20:20:20.358227916Z"
                differenceInMilliseconds={3600000}
                translateX={0}
                translateY={0}
                size={10}
                minTimeRange={0}
                maxTimeRange={3600000 * 2}
            />
        </svg>
    );
    expect(screen.getByTestId('policy-violation-event')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
});

test('should show a process activity event marker', async () => {
    const { asFragment } = render(
        <svg height={100} width={100} data-testid="timeline-main-view">
            <EventMarker
                id="1"
                name="eventName"
                type="ProcessActivityEvent"
                timestamp="2020-04-20T20:20:20.358227916Z"
                args="-g daemon off;"
                parentName={null}
                parentUid={-1}
                uid={1000}
                differenceInMilliseconds={3600000}
                translateX={0}
                translateY={0}
                size={10}
                minTimeRange={0}
                maxTimeRange={3600000 * 2}
            />
        </svg>
    );
    expect(screen.getByTestId('process-activity-event')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
});

test('should show a process in baseline activity event marker', async () => {
    const { asFragment } = render(
        <svg height={100} width={100} data-testid="timeline-main-view">
            <EventMarker
                id="1"
                name="eventName"
                type="ProcessActivityEvent"
                timestamp="2020-04-20T20:20:20.358227916Z"
                args="-g daemon off;"
                parentName={null}
                parentUid={-1}
                uid={1000}
                differenceInMilliseconds={3600000}
                inBaseline
                translateX={0}
                translateY={0}
                size={10}
                minTimeRange={0}
                maxTimeRange={3600000 * 2}
            />
        </svg>
    );
    expect(screen.getByTestId('process-in-baseline-activity-event')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
});

test('should show a container restart event marker', async () => {
    const { asFragment } = render(
        <svg height={100} width={100} data-testid="timeline-main-view">
            <EventMarker
                id="1"
                name="eventName"
                type="ContainerRestartEvent"
                timestamp="2020-04-20T20:20:20.358227916Z"
                differenceInMilliseconds={3600000}
                translateX={0}
                translateY={0}
                size={10}
                minTimeRange={0}
                maxTimeRange={3600000 * 2}
            />
        </svg>
    );
    expect(screen.getByTestId('restart-event')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
});

test('should show a container termination event marker', async () => {
    const { asFragment } = render(
        <svg height={100} width={100} data-testid="timeline-main-view">
            <EventMarker
                id="1"
                name="eventName"
                type="ContainerTerminationEvent"
                timestamp="2020-04-20T20:20:20.358227916Z"
                reason="OOMKilled"
                differenceInMilliseconds={3600000}
                translateX={0}
                translateY={0}
                size={10}
                minTimeRange={0}
                maxTimeRange={3600000 * 2}
            />
        </svg>
    );
    expect(screen.getByTestId('termination-event')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
});
