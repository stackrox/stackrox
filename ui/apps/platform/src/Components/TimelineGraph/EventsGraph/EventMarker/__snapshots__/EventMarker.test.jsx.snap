// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show a container restart event marker 1`] = `
<DocumentFragment>
  <svg
    data-testid="timeline-main-view"
    height="100"
    width="100"
  >
    <g
      data-testid="timeline-event-marker"
      transform="translate(NaN, -5)"
    >
      <g>
        <svg
          clip-rule="evenodd"
          data-testid="restart-event"
          fill-rule="evenodd"
          height="10"
          stroke-linejoin="round"
          stroke-miterlimit="2"
          viewBox="0 0 16 15"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.202 1.183l6.612 11.268a1.354 1.354 0 01-.508 1.866c-.211.12-.451.183-.694.183H1.388A1.377 1.377 0 010 13.134c0-.24.064-.475.186-.683L6.798 1.183a1.4 1.4 0 012.404 0z"
            fill="#ff9064"
          />
        </svg>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;

exports[`should show a container termination event marker 1`] = `
<DocumentFragment>
  <svg
    data-testid="timeline-main-view"
    height="100"
    width="100"
  >
    <g
      data-testid="timeline-event-marker"
      transform="translate(NaN, -5)"
    >
      <g>
        <svg
          data-testid="termination-event"
          height="10"
          viewBox="0 0 16 16"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.202 13.817l6.612-11.268a1.354 1.354 0 00-.508-1.866A1.406 1.406 0 0014.612.5H1.388C.621.5 0 1.112 0 1.866c0 .24.064.475.186.683l6.612 11.268a1.4 1.4 0 002.404 0z"
            fill="#FF9064"
            fill-rule="evenodd"
          />
        </svg>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;

exports[`should show a policy violation event marker 1`] = `
<DocumentFragment>
  <svg
    data-testid="timeline-main-view"
    height="100"
    width="100"
  >
    <g
      data-testid="timeline-event-marker"
      transform="translate(NaN, -5)"
    >
      <g>
        <svg
          data-testid="policy-violation-event"
          height="10"
          version="1.1"
          viewBox="0 0 15 15"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            transform="translate(0 .1)"
          >
            <rect
              fill="#FF5782"
              height="14.58"
              rx="2.43"
              width="14.58"
            />
            <path
              d="M8.071 8.37H6.616l-.325-5.584h2.12L8.07 8.37zm-1.833 2.22c0-.346.091-.614.275-.804.184-.19.458-.286.82-.286.364 0 .639.094.825.282.186.189.279.458.279.808 0 .345-.096.613-.29.804-.192.19-.463.285-.813.285-.359 0-.631-.096-.817-.289-.186-.192-.28-.46-.28-.8z"
              fill="#FFF"
            />
          </g>
        </svg>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;

exports[`should show a process activity event marker 1`] = `
<DocumentFragment>
  <svg
    data-testid="timeline-main-view"
    height="100"
    width="100"
  >
    <g
      data-testid="timeline-event-marker"
      transform="translate(NaN, -5)"
    >
      <g>
        <svg
          data-testid="process-activity-event"
          height="10"
          viewBox="0 0 16 16"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="#5677DD"
            fill-rule="evenodd"
            height="14.58"
            rx="2.43"
            transform="translate(-689 -673)"
            width="14.58"
            x="689"
            y="673.1"
          />
        </svg>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;

exports[`should show a process in baseline activity event marker 1`] = `
<DocumentFragment>
  <svg
    data-testid="timeline-main-view"
    height="100"
    width="100"
  >
    <g
      data-testid="timeline-event-marker"
      transform="translate(NaN, -5)"
    >
      <g>
        <svg
          data-testid="process-in-baseline-activity-event"
          height="10"
          viewBox="0 0 16 16"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="none"
            fill-rule="evenodd"
            transform="translate(0 .1)"
          >
            <rect
              fill="#56DDB2"
              height="14.58"
              rx="2.43"
              width="14.58"
            />
            <path
              d="M4.459 6.768a.807.807 0 00-1.18-.072.91.91 0 00-.067 1.24l2.685 3.17a.81.81 0 001.281-.043l4.645-6.244a.909.909 0 00-.144-1.233.808.808 0 00-1.172.151L6.48 9.153l-2.02-2.385z"
              fill="#FFF"
              fill-rule="nonzero"
            />
          </g>
        </svg>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;
